column_computation.hh

```diff
@@ -13,6 +13,12 @@
 class schema;
 class partition_key;
 class clustering_row;
+struct atomic_cell_view;
+struct tombstone;
+
+namespace db::view {
+struct bytes_with_action;
+}
```

Comment:
In the previous incarnation of this series we had a long discussion about why I didn't like bytes_with_action. What happened to it?

Reply:
Nothing. I took @enedil 's patches as-is (just fixing a few small conflicts and problems) and added specific patches at the end. If you didn't like bytes_with_action, your comments probably still apply and I can take a look at them.

Can you please either copy the relevant copy from that PR to here, or, alternatively, go to the old PR and "resolve" your comments which were either addressed or you don't insist on any more? Because I understood from @psarna that for some of the things you wanted to redesign, the decision was not to do that redesign. If there are things that still need to be redesigned before committing this series, I want to understand what those are. Thanks.

Reply:
I forgot the details, maybe @psarna remembers. If not I can try to recreate them.

Reply:
The original review was https://github.com/scylladb/scylla/pull/10123 (as I noted, it is not possible to change the ownership of a PR, which is why I was forced to create this new one). Maybe you can take a look at that and just "resolve" your comments that you don't care about any more, or were handled, and then I'll go there and also address your old comments that still need addressing.

Reply:
I see your comments about bytes_with_action in https://github.com/scylladb/scylla/pull/10123/commits/4389d2e107db8c80b6930898c7097c385022d167 but I didn't understand what exactly your complaint is :-( Since I also found this class hard to understand, I can try to improve it the way I see fit - but I can't be sure that I'll improve it in the same way that you want to improve it, so it will be good if you can tell me more about what exactly you didn't like.

Reply:
A computed column is a pure computation where the computed column's value is derived from some other columns. But that's not the case for bytes_with_actions - each individual "action" gets distributed into a new row. So it's abusing the column computation abstraction.

I don't remember if bytes_with_action violates the pureness property or not.

Reply:
My conclusion about the bytes_with_action discussion was posted here: https://github.com/scylladb/scylla/pull/10123#issuecomment-1160101326

tl;dr: one of the discussed solutions for this was to rename "column_computation" to something else, now that it starts serving a slightly different purpose, producing multiple rows. Technically I still think that the name "column computation" doesn't really imply a single value, because a column is a subset of values for given table, but that's bikeshedding, so I'm also absolutely fine with figuring out a different, more descriptive name. I'm not convinced by the alternative suggested here: https://github.com/scylladb/scylla/pull/10123#issuecomment-1157529922, for more context I recommend reading my original comment above.

Reply:
Column computation doesn't imply a single value, but "computed column" implies storing that value (perhaps of collection type) once, not generating multiple rows.

If we're renaming, let's also rename bytes_with_actions. It's incredibly non-descriptive. What are those bytes? and what are the actions?

Reply:
I still don't understand how renaming or refactoring or re-explaining **bytes_with_action** is related to renaming "column_computation"  or changing its return value from a vector of whatever, to something else.

I'll focus my efforts now on renaming or cleaning up bytes_with_actions itself - let me know if you want me to do something else.

Reply:
The more I look at this collection_column_computation code, the less I think it should be generic. The base class column_computation continues to support only a single return value - not a list. The subclass collection_column_computation has a new method returning a list - and it is only called in one place, in the view.cc code. That place actually use dynamic_cast to give collection_column_computation very special treatment - that no other future column_computation can get if it also wants to return multiple results.

I'm really tempted moving all this collection_column_computation code into a separate source file that lives in db/view/ and stop pretending it is something general. 

But before I do that, @avikivity @psarna I want to understand if one of you would object, and actually wanted the feature of computed column returning multiple values to be a generic feature that later we can reuse without change. If the answer is yes, I think the way it is written now is not general enough.

Reply:
Pardon me, I somehow missed this comment 2 days ago - I also agree that making things generic without a very good reason is a mistake, so moving this code to a separate file with specialized implementation sounds good. And as for naming bytes_with_action, I don't have any good suggestions, but here's a mediocre one: `value_or_tombstone_or_marker`

Reply:
Sadly, the more I look at it, the less sense bytes_with_actions makes. get_view_rows() actually returns not just bytes_with_action, but a row *and* an action. So it views bytes_with_action as an additional action, not the main row addition, but in the collection column indexing, there is *only* this bytes_with_action.

I'm trying now to write comments around this, but this organization is very hard to explain or defend. On the other hand, I don't see what other alternative I have besides spending many days rewriting this already-working code.